File: Algorithm/Baekjoon/17298.cpp
#include <iostream>
#include <stack>
#define MAX 1000001

using namespace std;

int n;
int seq[MAX];
int answer[MAX];

void solution() {
    stack<int> stack;

    for (int i = n - 1; i >= 0; --i) {
        while (!stack.empty() && stack.top() <= seq[i]) {
            stack.pop();
        }

        if (stack.empty()) {
            answer[i] = -1;
        } else {
            answer[i] = stack.top();
        }

        stack.push(seq[i]);
    }

    for (int i = 0; i < n; ++i) {
        cout << answer[i] << " ";
    }
}

int main() {
    cin >> n;
    for (int i = 0; i < n; ++i) {
        cin >> seq[i];
    }

    solution();

    return 0;
}

File: Algorithm/Baekjoon/17298.swift
import Foundation

let n = Int(readLine()!)!
let seq = readLine()!.split(separator: " ").map { Int($0)! }

var stack = Stack()
var answer = [Int](repeating: 0, count: n)

for i in stride(from: n - 1, through: 0, by: -1) {
    while !stack.isEmpty && stack.top <= seq[i] {
        _ = stack.pop()
    }

    if stack.isEmpty {
        answer[i] = -1
    } else {
        answer[i] = stack.top
    }

    stack.push(seq[i])
}

for num in answer {
    print(num, terminator: " ")
}

struct Stack {

    var stack = [Int]()

    var count: Int { stack.count }
    var isEmpty: Bool { stack.isEmpty }
    var top: Int { stack.last! }

    mutating func push(_ element: Int) {
        stack.append(element)
    }

    mutating func pop() -> Int {
        return stack.removeLast()
    }
}